### PR TITLE
widgets: algin rect outline radius with inner border radius

### LINKED ISSUE #611: Shape borders don't align with rounded corners
### Regression?

No

### Hyprlock Info and Version

<Enter your hyprlock version here>Hyprlock version 0.6.0
<details> <summary>Hyprlock config</summary>

```sh
shape {
  monitor =

  border_size = 200
  color = rgb(ff0000)

  size = 900, 1300
  rounding = 150

  position = 0, 0
  halign = center
  valign = center
}

```

</details>

### Compositor Info and Version

<details> <summary>System/Version info</summary>

```sh
Hyprland 0.46.2 built from branch  at commit 0bd541f2fd902dbfa04c3ea2ccf679395e316887  (version: bump to 0.46.2).
Date: Thu Dec 19 19:26:47 2024
Tag: v0.46.2, commits: 5566
built against:
 aquamarine 0.5.1
 hyprlang 0.6.0
 hyprutils 0.3.0
 hyprcursor 0.1.11
 hyprgraphics 0.1.1


flags set:
debug


System Information:
System name: Linux
Node name: cypher
Release: 6.12.6-arch1-1
Version: #1 SMP PREEMPT_DYNAMIC Thu, 19 Dec 2024 21:29:01 +0000


GPU information: 
c1:00.0 VGA compatible controller [0300]: Advanced Micro Devices, Inc. [AMD/ATI] Phoenix1 [1002:15bf] (rev c2) (prog-if 00 [VGA controller])


os-release: NAME="Arch Linux"
PRETTY_NAME="Arch Linux"
ID=arch
BUILD_ID=rolling
ANSI_COLOR="38;2;23;147;209"
HOME_URL="https://archlinux.org/"
DOCUMENTATION_URL="https://wiki.archlinux.org/"
SUPPORT_URL="https://bbs.archlinux.org/"
BUG_REPORT_URL="https://gitlab.archlinux.org/groups/archlinux/-/issues"
PRIVACY_POLICY_URL="https://terms.archlinux.org/docs/privacy-policy/"
LOGO=archlinux-logo


plugins:

```

</details>

### Description

I made a super ugly config to show case the issue. The shape's inner border radius doesn't follow the shape's rounding radius. I don't think this is intended. In my example screenshot, I'd expect the light blue part to touch the red part.

### How to reproduce

I provided a minimal example above that produces the problematic shape.

### Crash reports, logs, images, videos

ugly borders

### PR DESCRIPTION
**Pull request by PaideiaDilemma**
Closes #611

To verify that the the outline of a box matches with the inner border perfectly,
I used this:

```
input-field {
  monitor=
  size=25%, 5%
  inner_color=rgba(fff7ed20)
  outer_color=rgba(fff7ed20)
  position=0, -20
  rounding=15
  fade_on_empty = false
  valign=center
}
```

Because inner and outer color have the same alpha, one can easily spot if they overlap or not.

Previously if the rounding was to big, it used rounding 0.
Now it clamps it to the maximum.

Tested:
- [x] rouding -1 for shape, image and input-field
- [x] rouding -1 and border for shape, image and input-field
- [x] rounding 0 for shape, image and input-field
- [x] rounding 0 and border for shape, image and input-field

### CODE REVIEW

**Comment by PaideiaDilemma** [comment]
This is what we want i think.
The inner radius of the border just needed to be scaled properly. No need for `outerRadius` in the shader.

Actually the problem just occurred in edge cases, where the rounding approached minhalf and a thick border was used.

**Comment by PaideiaDilemma** [comment]
This is what we want i think.
The inner radius of the border just needed to be scaled properly. No need for `outerRadius` in the shader.

Actually the problem just occurred in edge cases, where the rounding approached minhalf and a thick border was used.

File: src/renderer/widgets/IWidget.hpp [hunk 1]
```diff
@@ -23,5 +25,5 @@ class IWidget {
         bool        allowForceUpdate = false;
     };
 
-    virtual SFormatResult formatString(std::string in);
+    static SFormatResult formatString(std::string in);
```
Comment (vaxerski) [inline src/renderer/widgets/IWidget.hpp:28]: should this be static?

File: src/renderer/widgets/IWidget.hpp [hunk 2]
```diff
@@ -10,10 +10,12 @@ class IWidget {
     };
     virtual ~IWidget() = default;
 
-    virtual bool     draw(const SRenderData& data) = 0;
+    virtual bool    draw(const SRenderData& data) = 0;
 
-    virtual Vector2D posFromHVAlign(const Vector2D& viewport, const Vector2D& size, const Vector2D& offset, const std::string& halign, const std::string& valign,
-                                    const double& ang = 0);
+    static Vector2D posFromHVAlign(const Vector2D& viewport, const Vector2D& size, const Vector2D& offset, const std::string& halign, const std::string& valign,
```
Comment (vaxerski) [inline src/renderer/widgets/IWidget.hpp:15]: same here

File: src/renderer/widgets/IWidget.hpp (hunk 1 above)
Comment (PaideiaDilemma) [reply src/renderer/widgets/IWidget.hpp:28]: I think so, because they are never overridden and they don't need an object.

**Review by vaxerski — APPROVED** [review]

**Pull request merged**